V3.0rc1 has runtime dep. on setuptools

### bug描述 Describe the Bug

Look here: https://github.com/PaddlePaddle/Paddle/blob/91a62d6879cb1df56ab6aced79ce4a6b341f6c99/python/paddle/utils/cpp_extension/cpp_extension.py#L23

I am getting the following error in CI:

```text
kreuzberg\_ocr\_paddleocr.py:242: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv\Lib\site-packages\paddleocr\__init__.py:14: in <module>
    from .paddleocr import (
.venv\Lib\site-packages\paddleocr\paddleocr.py:21: in <module>
    from paddle.utils import try_import
.venv\Lib\site-packages\paddle\__init__.py:38: in <module>
    from .base import core  # noqa: F401
.venv\Lib\site-packages\paddle\base\__init__.py:38: in <module>
    from . import (  # noqa: F401
.venv\Lib\site-packages\paddle\base\dataset.py:23: in <module>
    from ..utils import deprecated
.venv\Lib\site-packages\paddle\utils\__init__.py:16: in <module>
    from . import (  # noqa: F401
.venv\Lib\site-packages\paddle\utils\cpp_extension\__init__.py:15: in <module>
    from .cpp_extension import (
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    # Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
    #
    # Licensed under the Apache License, Version 2.0 (the "License");
    # you may not use this file except in compliance with the License.
    # You may obtain a copy of the License at
    #
    #     http://www.apache.org/licenses/LICENSE-2.0
    #
    # Unless required by applicable law or agreed to in writing, software
    # distributed under the License is distributed on an "AS IS" BASIS,
    # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    # See the License for the specific language governing permissions and
    # limitations under the License.
    
    # isort: skip_file
    
    from __future__ import annotations
    from typing import TYPE_CHECKING, Any
    import os
    import copy
    import concurrent
    import re
>   import setuptools
E   ModuleNotFoundError: No module named 'setuptools'

.venv\Lib\site-packages\paddle\utils\cpp_extension\cpp_extension.py:23: ModuleNotFoundError

The above exception was the direct cause of the following exception:

backend = <kreuzberg._ocr._paddleocr.PaddleBackend object at 0x000001DE22318290>
mock_find_spec = <MagicMock name='find_spec' id='205356[805](https://github.com/Goldziher/kreuzberg/actions/runs/13844252481/job/38738917361?pr=27#step:11:806)0640'>

    @pytest.mark.anyio
    @pytest.mark.skipif(sys.version_info > (3, 12), reason="paddle does not support python 3.13 yet")
    async def test_init_paddle_ocr_initialization_error(backend: PaddleBackend, mock_find_spec: Mock) -> None:
        """Test handling initialization errors."""
    
        PaddleBackend._paddle_ocr = None
    
        async def mock_run_sync_error(*args: Any, **_: Any) -> None:
            if args and args[0].__name__ == "PaddleOCR":
                raise Exception("Initialization error")
    
        with patch("kreuzberg._ocr._paddleocr.run_sync", side_effect=mock_run_sync_error):
            with pytest.raises(OCRError) as excinfo:
>               await backend._init_paddle_ocr()

tests\ocr\paddleocr_test.py:234: 
```

### 其他补充信息 Additional Supplementary Information

_No response_

--- a/python/requirements.txt
+++ b/python/requirements.txt
@@ -6,3 +6,4 @@ opt_einsum==3.3.0
 networkx
 typing_extensions
 safetensors>=0.6.0
+setuptools; python_version>="3.12"

--- a/python/setup.py.in
+++ b/python/setup.py.in
@@ -1637,7 +1637,7 @@ def check_build_dependency():
 Please run 'pip install -r python/requirements.txt' to make sure you have all the dependencies installed.
 '''.strip()
 
-    with open(TOP_DIR + '/python/requirements.txt') as f:
+    with open('${PADDLE_SOURCE_DIR}' + '/python/requirements.txt') as f:
         build_dependencies = (
             f.read().splitlines()
         )  # Specify the dependencies to install
@@ -1654,22 +1654,33 @@ Please run 'pip install -r python/requirements.txt' to make sure you have all th
 
         # Build environment dict
         env_markers = {
-            'python_version': f"{sys.version_info.major}.{sys.version_info.minor}",
-            'python_full_version': f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}",
-            'platform_system': platform.system(),
-            'platform_machine': platform.machine(),
-            'sys_platform': sys.platform,
+            'python_version': (sys.version_info.major, sys.version_info.minor),
+            'python_full_version': (
+                sys.version_info.major,
+                sys.version_info.minor,
+                sys.version_info.micro,
+            ),
+            'platform_system': f'"{platform.system()}"',
+            'platform_machine': f'"{platform.machine()}"',
+            'sys_platform': f'"{sys.platform}"',
         }
 
-        # Simple marker evaluation - handle common cases
-        # This is a simplified version that handles the most common patterns
+        # Marker evaluation
         try:
-            # Replace marker variables with their values
             eval_str = marker_str
+            # Replace marker variables with their values
             for key, value in env_markers.items():
-                eval_str = eval_str.replace(key, f'"{value}"')
+                eval_str = eval_str.replace(key, str(value))
 
-            # Evaluate the expression
+            def version_to_tuple(match):
+                version_str = match.group(1)
+                parts = version_str.split('.')
+                return '(' + ', '.join(parts) + ')'
+
+            eval_str = re.sub(
+                r'["\'](\d+(?:\.\d+)*)["\']', version_to_tuple, eval_str
+            )
+
             return eval(eval_str)
         except Exception as e:
             raise RuntimeError(f"Failed to evaluate marker '{marker_str}': {e}")

--- a/python/setup.py.in
+++ b/python/setup.py.in
@@ -1637,7 +1637,7 @@ def check_build_dependency():
 Please run 'pip install -r python/requirements.txt' to make sure you have all the dependencies installed.
 '''.strip()
 
-    with open('${PADDLE_SOURCE_DIR}' + '/python/requirements.txt') as f:
+    with open(TOP_DIR + '/python/requirements.txt') as f:
         build_dependencies = (
             f.read().splitlines()
         )  # Specify the dependencies to install
@@ -1645,10 +1645,60 @@ Please run 'pip install -r python/requirements.txt' to make sure you have all th
     python_dependencies_module = []
     installed_packages = []
 
+    def eval_marker(marker_str):
+        """Simple evaluation of PEP 508 environment markers."""
+        if not marker_str:
+            return True
+
+        marker_str = marker_str.strip()
+
+        # Build environment dict
+        env_markers = {
+            'python_version': f"{sys.version_info.major}.{sys.version_info.minor}",
+            'python_full_version': f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}",
+            'platform_system': platform.system(),
+            'platform_machine': platform.machine(),
+            'sys_platform': sys.platform,
+        }
+
+        # Simple marker evaluation - handle common cases
+        # This is a simplified version that handles the most common patterns
+        try:
+            # Replace marker variables with their values
+            eval_str = marker_str
+            for key, value in env_markers.items():
+                eval_str = eval_str.replace(key, f'"{value}"')
+
+            # Evaluate the expression
+            return eval(eval_str)
+        except Exception as e:
+            raise RuntimeError(f"Failed to evaluate marker '{marker_str}': {e}")
+
     for dependency in build_dependencies:
-        python_dependencies_module.append(
-            re.sub("_|-", '', re.sub(r"==.*|>=.*|<=.*", '', dependency))
-        )
+        dependency = dependency.strip()
+        if not dependency or dependency.startswith('#'):
+            continue
+
+        # Split dependency spec and environment marker
+        if ';' in dependency:
+            dependency_spec, marker = dependency.split(';', 1)
+            dependency_spec = dependency_spec.strip()
+            marker = marker.strip()
+
+            # Evaluate marker - skip if not applicable to current environment
+            if not eval_marker(marker):
+                continue
+        else:
+            dependency_spec = dependency
+
+        # Remove version specifiers from dependency spec
+        dependency_name = re.sub(
+            r"==.*|>=.*|<=.*|~=.*|!=.*", '', dependency_spec
+        ).strip()
+
+        # Normalize package name (remove _ and -)
+        python_dependencies_module.append(re.sub("_|-", '', dependency_name))
+
     reqs = subprocess.check_output([sys.executable, '-m', 'pip', 'freeze'])
 
     for r in reqs.split():

--- a/setup.py
+++ b/setup.py
@@ -2576,22 +2576,33 @@ Please run 'pip install -r python/requirements.txt' to make sure you have all th
 
         # Build environment dict
         env_markers = {
-            'python_version': f"{sys.version_info.major}.{sys.version_info.minor}",
-            'python_full_version': f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}",
-            'platform_system': platform.system(),
-            'platform_machine': platform.machine(),
-            'sys_platform': sys.platform,
+            'python_version': (sys.version_info.major, sys.version_info.minor),
+            'python_full_version': (
+                sys.version_info.major,
+                sys.version_info.minor,
+                sys.version_info.micro,
+            ),
+            'platform_system': f'"{platform.system()}"',
+            'platform_machine': f'"{platform.machine()}"',
+            'sys_platform': f'"{sys.platform}"',
         }
 
-        # Simple marker evaluation - handle common cases
-        # This is a simplified version that handles the most common patterns
+        # Marker evaluation
         try:
-            # Replace marker variables with their values
             eval_str = marker_str
+            # Replace marker variables with their values
             for key, value in env_markers.items():
-                eval_str = eval_str.replace(key, f'"{value}"')
+                eval_str = eval_str.replace(key, str(value))
 
-            # Evaluate the expression
+            def version_to_tuple(match):
+                version_str = match.group(1)
+                parts = version_str.split('.')
+                return '(' + ', '.join(parts) + ')'
+
+            eval_str = re.sub(
+                r'["\'](\d+(?:\.\d+)*)["\']', version_to_tuple, eval_str
+            )
+
             return eval(eval_str)
         except Exception as e:
             raise RuntimeError(f"Failed to evaluate marker '{marker_str}': {e}")

--- a/setup.py
+++ b/setup.py
@@ -2567,10 +2567,60 @@ Please run 'pip install -r python/requirements.txt' to make sure you have all th
     python_dependencies_module = []
     installed_packages = []
 
+    def eval_marker(marker_str):
+        """Simple evaluation of PEP 508 environment markers."""
+        if not marker_str:
+            return True
+
+        marker_str = marker_str.strip()
+
+        # Build environment dict
+        env_markers = {
+            'python_version': f"{sys.version_info.major}.{sys.version_info.minor}",
+            'python_full_version': f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}",
+            'platform_system': platform.system(),
+            'platform_machine': platform.machine(),
+            'sys_platform': sys.platform,
+        }
+
+        # Simple marker evaluation - handle common cases
+        # This is a simplified version that handles the most common patterns
+        try:
+            # Replace marker variables with their values
+            eval_str = marker_str
+            for key, value in env_markers.items():
+                eval_str = eval_str.replace(key, f'"{value}"')
+
+            # Evaluate the expression
+            return eval(eval_str)
+        except Exception as e:
+            raise RuntimeError(f"Failed to evaluate marker '{marker_str}': {e}")
+
     for dependency in build_dependencies:
-        python_dependencies_module.append(
-            re.sub("_|-", '', re.sub(r"==.*|>=.*|<=.*", '', dependency))
-        )
+        dependency = dependency.strip()
+        if not dependency or dependency.startswith('#'):
+            continue
+
+        # Split dependency spec and environment marker
+        if ';' in dependency:
+            dependency_spec, marker = dependency.split(';', 1)
+            dependency_spec = dependency_spec.strip()
+            marker = marker.strip()
+
+            # Evaluate marker - skip if not applicable to current environment
+            if not eval_marker(marker):
+                continue
+        else:
+            dependency_spec = dependency
+
+        # Remove version specifiers from dependency spec
+        dependency_name = re.sub(
+            r"==.*|>=.*|<=.*|~=.*|!=.*", '', dependency_spec
+        ).strip()
+
+        # Normalize package name (remove _ and -)
+        python_dependencies_module.append(re.sub("_|-", '', dependency_name))
+
     reqs = subprocess.check_output([sys.executable, '-m', 'pip', 'freeze'])
 
     for r in reqs.split():

